Extension/src/LanguageServer/settings.ts

```diff
@@ -452,6 +461,224 @@ export class CppSettings extends Settings {
+            edits.createFile(uri, { ignoreIfExists: true, overwrite: false });
+            try {
+                await vscode.workspace.applyEdit(edits);
+                document = await vscode.workspace.openTextDocument(uri);
+            } catch (e) {
+                document = await vscode.workspace.openTextDocument();
+            }
+        } else {
+            document = await vscode.workspace.openTextDocument();
+        }
+        this.populateEditorConfig(document);
+    }
+
+    // If formattingEngine is set to "Default", searches for .editorconfig with vcFormat
+    // entries, or a .clang-format file, to determine which settings to use.
+    // This is intentionally not async to avoid races due to multiple entrancy.
+    public useVcFormat(document: vscode.TextDocument): boolean {
+        if (this.formattingEngine !== "Default") {
+            return this.formattingEngine === "vcFormat";
+        }
+        if (this.clangFormatStyle !== "file") {
+            // If a clang-format style other than file is specified, don't try to switch to vcFormat.
+            return false;
+        }
+        const cachedValue: boolean | undefined = cachedEditorConfigLookups.get(document.uri.fsPath);
+        if (cachedValue !== undefined) {
+            return cachedValue;
+        }
+        let foundEditorConfigWithVcFormatSettings: boolean = false;
+        const findConfigFile: (parentPath: string) => boolean = (parentPath: string) => {
+            const editorConfigPath: string = path.join(parentPath, ".editorconfig");
+            if (fs.existsSync(editorConfigPath)) {
+                const editorConfigSettings: any = getEditorConfigSettings(document.uri.fsPath);
+                const keys: string[] = Object.keys(editorConfigSettings);
+                for (let i: number = 0; i < keys.length; ++i) {
+                    if (keys[i].startsWith("cpp_")) {
+                        foundEditorConfigWithVcFormatSettings = true;
+                        const didEditorConfigNotice: PersistentState<boolean> = new PersistentState<boolean>("Cpp.didEditorConfigNotice", false);
+                        if (!didEditorConfigNotice.Value) {
+                            vscode.window.showInformationMessage(localize("editorconfig.default.behavior",
+                                "vcFormat was selected because an '.editorconfig' file was found containing vcFormat entries and 'Cpp.formatting' is set to 'Default'."));
+                            didEditorConfigNotice.Value = true;
+                        }
+                        return true;
+                    }
+                }
+                if (editorConfigSettings.root?.toLowerCase() === "true") {
+                    return true;
+                }
+            } else {
+                const clangFormatPath1: string = path.join(parentPath, ".clang-format");
+                if (fs.existsSync(clangFormatPath1)) {
+                    return true;
```

Comment:
Should this be false?  Based on how it's used in the do..while, I thought you expected this to only return true for .editorconfig found.

Reply:
I'm returning true from the lambda to indicate that a file was found (either `.clang-format` or `.editorconfig`).  Which one found is specified by `foundEditorConfigWithVcFormatSettings`.  If false is returned, it moves up to check the next parent directory.